components/file_system/src/iosnoop/biosnoop.rs

```diff
@@ -0,0 +1,332 @@
+// Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.
+
+use super::metrics::*;
+use super::IOStats;
+use crate::IOType;
+
+use collections::HashMap;
+use std::collections::VecDeque;
+use std::ffi::CString;
+use std::ptr;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::Mutex;
+
+use bcc::{table::Table, Kprobe, BPF};
+use crossbeam_utils::CachePadded;
+
+const MAX_THREAD_IDX: usize = 128;
+
+static mut BPF_TABLE: Option<(BPF, Table, Table)> = None;
+// This array records the io type for every thread. The address of this array
+// will be passed into BPF, so BPF code can get io type for specific thread
+// without an extra syscall.
+// It should be a thread local variable, but the address of thread local is not
+// reliable. So define a global array and let each thread writes on a specific
+// element. Thus there is no contention for the element and use padding to avoid
+// false sharing.
+static mut IO_TYPE_ARRAY: [CachePadded<IOType>; MAX_THREAD_IDX] =
```

Comment:
Does the inner IOType needs to be atomic?

Reply:
no, every thread only accesses the element of its own index, so no contention between threads.